### PR TITLE
Having the payload generator check the payload size

### LINKED ISSUE #6558: msfvenom -s option doesn't work
Hi, it looks like msfvenom is ignoring -s option if no encoder is specified:

```
# msfvenom -a x86 --platform Windows -p windows/meterpreter/reverse_http LHOST=IP LPORT=80 -s 800 -e x86/alpha_mixed -i 3 -f raw -o rev_http_IP_80
Found 1 compatible encoders
Attempting to encode payload with 3 iterations of x86/alpha_mixed
x86/alpha_mixed succeeded with size 705 (iteration=0)
x86/alpha_mixed succeeded with size 1472 (iteration=1)
x86/alpha_mixed failed with encoder has made a buffer that is too big
Error: An encoding exception occurred.

# msfvenom -a x86 --platform Windows -p windows/meterpreter/reverse_http LHOST=IP LPORT=80 -s 100 -f raw -o rev_http_IP_80
No encoder or badchars specified, outputting raw payload
Payload size: 322 bytes
Saved as: rev_http_IP_80

### PR DESCRIPTION
**Pull request by darkbushido**
Checking the size of the payload and raising an error if its to larger then the size option. This should fix #6558
## Verification

List the steps needed to make sure this thing works
- [x] `msfvenom -a x86 --platform Windows -p windows/meterpreter/reverse_http LHOST=IP LPORT=80 -s 100 -f raw -o rev_http_IP_80` should not generate a payload

### CODE REVIEW

**Comment by wvu-r7** [comment]
Please do not merge yet. Working on the spec.

**Comment by wvu-r7** [comment]
@darkbushido: This is so much better now. Thanks for rewriting the spec.

**Comment by wvu-r7** [comment]
Hmm, it's clearer now, but there's a lot of code duplication. We'll have to live with it for now.

**Comment by bcook-r7** [comment]
I prefer the duplication. It's now clear what everything is doing :)

**Comment by darkbushido** [comment]
I think i can cut down on the duplication, or at least fix the nesting so it makes sense.

File: lib/msf/core/payload_generator.rb [hunk 1]
```diff
@@ -314,20 +317,25 @@ def generate_payload
...
       else
         raw_payload = generate_raw_payload
         raw_payload = add_shellcode(raw_payload)
         encoded_payload = encode_payload(raw_payload)
         encoded_payload = prepend_nops(encoded_payload)
         cli_print "Payload size: #{encoded_payload.length} bytes"
-        format_payload(encoded_payload)
+        gen_payload = format_payload(encoded_payload)
+      end
+      if gen_payload.length > @space and not @smallest
```
Comment (busterb) [inline lib/msf/core/payload_generator.rb]: this errors out in an uncaught way if gen_payload is nil. That happens when an issue like #6915 occurs (wrong format for the payload).

Suggest a patch like this and kill 2 birds with one stone:
```
diff --git a/lib/msf/core/payload_generator.rb b/lib/msf/core/payload_generator.rb
index 987871c..ec8c96b 100644
--- a/lib/msf/core/payload_generator.rb
+++ b/lib/msf/core/payload_generator.rb
@@ -332,7 +332,9 @@ module Msf
         cli_print "Payload size: #{encoded_payload.length} bytes"
         gen_payload = format_payload(encoded_payload)
       end
-      if gen_payload.length > @space and not @smallest
+      if gen_payload.nil?
+        raise PayloadGeneratorError, 'The payload could not be generated, check options'
+      elsif gen_payload.length > @space and not @smallest
         raise PayloadSpaceViolation, 'The payload exceeds the specified space'
       else
         gen_payload
```

**Comment by bcook-r7** [comment]
I'm going to grab this, have an additional spec and fix to add.

**Comment by wvu-r7** [comment]
Thanks, @bcook-r7. FYI, https://github.com/rapid7/metasploit-framework/commit/96c459c71d1d8b4a5704819b412afeaa459c11fa and https://github.com/rapid7/metasploit-framework/commit/c2cf992560336a2377fac84971a990a5d8434001 were added.

**Comment by bcook-r7** [comment]
NP! definitely looks like we can trim out some of the redundant settings from some of the other tests as well in a follow-up.

**Pull request merged**